src/driver/result.rs

```diff
@@ -568,3 +568,108 @@ pub unsafe fn launch_kernel(
+    /// # Safety
+    /// `size` must be the size of the size of the memory object in bytes.
+    #[cfg(windows)]
+    pub unsafe fn import_external_memory_opaque_win32(
+        handle: std::os::windows::io::RawHandle,
+        size: u64,
+    ) -> Result<sys::CUexternalMemory, DriverError> {
+        let mut external_memory = MaybeUninit::uninit();
+        let handle_description = sys::CUDA_EXTERNAL_MEMORY_HANDLE_DESC {
+            type_: sys::CUexternalMemoryHandleType::CU_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_WIN32,
+            handle: sys::CUDA_EXTERNAL_MEMORY_HANDLE_DESC_st__bindgen_ty_1 {
+                win32: sys::CUDA_EXTERNAL_MEMORY_HANDLE_DESC_st__bindgen_ty_1__bindgen_ty_1 {
+                    handle,
+                    name: std::ptr::null(),
+                },
+            },
+            size,
+            ..Default::default()
+        };
+        sys::cuImportExternalMemory(external_memory.as_mut_ptr(), &handle_description).result()?;
+        Ok(external_memory.assume_init())
+    }
+
+    /// Destroys an external memory object.
+    ///
+    /// See [cuda docs](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__EXTRES__INTEROP.html#group__CUDA__EXTRES__INTEROP_1g1b586dda86565617e7e0883b956c7052)
+    ///
+    /// # Safety
+    /// - Any mapped buffers onto this object must already be freed.
+    /// - The external memory must only be destroyed once.
+    pub unsafe fn destroy_external_memory(
+        external_memory: sys::CUexternalMemory,
+    ) -> Result<(), DriverError> {
+        sys::cuDestroyExternalMemory(external_memory).result()
+    }
+
+    /// Maps a buffer onto an imported memory object.
+    ///
+    /// See [cuda docs](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__EXTRES__INTEROP.html#group__CUDA__EXTRES__INTEROP_1gb9fec33920400c70961b4e33d838da91)
+    ///
+    /// # Safety
+    /// Mapped buffers may overlap.
+    pub unsafe fn get_mapped_buffer(
+        external_memory: sys::CUexternalMemory,
+        offset: u64,
+        size: u64,
+    ) -> Result<sys::CUdeviceptr, DriverError> {
+        let mut device_ptr = MaybeUninit::uninit();
+        let buffer_description = sys::CUDA_EXTERNAL_MEMORY_BUFFER_DESC {
+            offset,
+            size,
+            ..Default::default()
+        };
+        sys::cuExternalMemoryGetMappedBuffer(
+            device_ptr.as_mut_ptr(),
+            external_memory,
+            &buffer_description,
+        )
+        .result()?;
+        Ok(device_ptr.assume_init())
+    }
+
+    /// Frees device memory.
+    ///
+    /// see [cuda docs](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEM.html#group__CUDA__MEM_1g89b3f154e17cc89b6eea277dbdf5c93a)
+    ///
+    /// # Safety
+    /// - Memory must only be freed once.
+    /// - All async accesses to this pointer must have been completed.
+    pub unsafe fn memory_free(device_ptr: sys::CUdeviceptr) -> Result<(), DriverError> {
+        sys::cuMemFree_v2(device_ptr).result()
+    }
```

Comment:
We have free_async a top level mod, want to move this up a level? It doesn't seem to be specific to the external memory module.

Reply:
Makes sense. Are you ok with the name `memory_free`?

Reply:
Yep!